constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -403,6 +425,100 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display> QuadraticSymbolicExp
+        &self,
+        range_constraints: &impl RangeConstraintProvider<T, V>,
+    ) -> Result<ProcessResult<T, V>, Error> {
+        let Some((left, right)) = self.try_as_single_product() else {
+            return Ok(ProcessResult::empty());
+        };
+        // Now we have `left * right = 0`, i.e. one (or both) of them has to be zero.
+        let (left_solution, right_solution) = match (
+            left.solve(range_constraints),
+            right.solve(range_constraints),
+        ) {
+            // If one of them is always unsatisfiable, it is equivalent to just solving the other one for zero.
+            (Err(_), o) | (o, Err(_)) => {
+                return o;
+            }
+            (Ok(left), Ok(right)) => (left, right),
+        };
+
+        if let Some(result) =
+            combine_to_conditional_assignment(left_solution, right_solution, range_constraints)
+        {
+            Ok(result)
+        } else {
+            // TODO Now at least try to combine joint range constraints
+            Ok(ProcessResult::empty())
+        }
+    }
+}
+
+/// Tries to combine two process results from alternative branches into a
+/// conditional assignment.
+fn combine_to_conditional_assignment<T: FieldElement, V: Ord + Clone + Hash + Eq + Display>(
+    left: ProcessResult<T, V>,
+    right: ProcessResult<T, V>,
+    range_constraints: &impl RangeConstraintProvider<T, V>,
+) -> Option<ProcessResult<T, V>> {
+    let [Effect::Assignment(first_var, first_assignment)]: [_; 1] = left.effects.try_into().ok()?
+    else {
+        return None;
+    };
+    let [Effect::Assignment(second_var, second_assignment)]: [_; 1] =
+        right.effects.try_into().ok()?
+    else {
+        return None;
+    };
+
+    if first_var != second_var {
+        return None;
+    }
+
+    // At this point, we have two assignments to the same variable, i.e.
+    // "`X = A` or `X = B`". If the two alternatives can never be satisfied at
+    // the same time (i.e. the "or" is exclusive), we can turn this into a
+    // conditional assignment.
+
+    let diff = symbolic_expression_to_quadratic_symbolic_expression(
+        &(first_assignment.clone() + -&second_assignment),
+    )?;
+    let diff = diff.try_to_known()?.try_to_number()?;
+    // `diff = A - B` is a compile-time known number, i.e. `A = B + diff`.
+    // Now if `rc + diff` is disjoint from `rc`, it means
+    // that if the value that `A` evaluates to falls into the allowed range for `X`,
+    // then `B = A + diff` is not a possible value for `X` and vice-versa.
+    // This means the two alternatives are disjoint and we can use a conditional assignment.
+    let rc = range_constraints.get(&first_var);
+    if !rc
+        .combine_sum(&RangeConstraint::from_value(diff))
+        .is_disjoint(&rc)
+    {
+        return None;
+    }
+
+    Some(ProcessResult {
+        effects: vec![Effect::ConditionalAssignment {
+            variable: first_var,
+            condition: BranchCondition {
```

Comment:
```suggestion
            condition: Condition {
```
Guess we should rename it?